OrbitService/LinuxTracingGrpcHandler.cpp

```diff
@@ -163,31 +172,47 @@ uint64_t LinuxTracingGrpcHandler::InternStringIfNecessaryAndGetKey(
 }
 
 void LinuxTracingGrpcHandler::SenderThread() {
-  constexpr std::chrono::duration kSendEvery = std::chrono::milliseconds{20};
-  std::chrono::time_point last_sent = std::chrono::steady_clock::now();
-  while (tracer_ != nullptr) {
-    std::chrono::duration since_sent =
-        std::chrono::steady_clock::now() - last_sent;
-    if (since_sent < kSendEvery) {
-      std::this_thread::sleep_for(kSendEvery - since_sent);
+  constexpr absl::Duration kSendTimeInterval = absl::Milliseconds(20);
+  // This should be lower than kMaxEventsPerResponse in SendBufferedEvents as
+  // a few more events are likely to arrive after the condition becomes true.
+  constexpr uint64_t kSendEventInterval = 5000;
+
+  while (true) {
+    event_buffer_mutex_.LockWhenWithTimeout(
+        absl::Condition(
+            +[](LinuxTracingGrpcHandler* self) {
+              return self->event_buffer_.size() >= kSendEventInterval ||
+                     self->tracer_ == nullptr;
+            },
+            this),
+        kSendTimeInterval);
+    if (tracer_ == nullptr) {
+      event_buffer_mutex_.Unlock();
+      break;
     }
-    last_sent = std::chrono::steady_clock::now();
-    SendBufferedEvents();
+    std::vector<CaptureEvent> buffered_events = std::move(event_buffer_);
+    event_buffer_.clear();
+    event_buffer_mutex_.Unlock();
+    SendBufferedEvents(std::move(buffered_events));
   }
-  SendBufferedEvents();
-}
 
-void LinuxTracingGrpcHandler::SendBufferedEvents() {
-  std::vector<CaptureEvent> events;
   {
-    absl::MutexLock lock{&event_buffer_mutex_};
-    events = std::move(event_buffer_);
+    event_buffer_mutex_.Lock();
+    std::vector<CaptureEvent> buffered_events = std::move(event_buffer_);
     event_buffer_.clear();
+    event_buffer_mutex_.Unlock();
+    SendBufferedEvents(std::move(buffered_events));
```

Comment:
Actually thinking about this more... since we now have tracer_.reset() protected by the mutex this part does not seem to be necessary anymore. With changes described above the queue should always be empty here.

Reply:
Good point, thanks, done, maybe have a final look at this.